WcaOnRails/app/models/competition.rb

```diff
@@ -96,25 +99,17 @@ def info
     info
   end
 
-  before_validation :clone_competition, on: [:create]
-  def clone_competition
-    if competition_id_to_clone.present?
-      competition_to_clone = Competition.find_by_id(competition_id_to_clone)
-      if competition_to_clone
-        attributes = competition_to_clone.attributes
-        # Don't clone id, name, cellName, and dates.
-        %w(id name cellName year month day endMonth endDay registration_open registration_close results_posted_at).each { |attribute| attributes.delete attribute }
-        # Make sure the new competition is not publicly visible and is open for
-        # editing.
-        attributes["showAtAll"] = false
-        attributes["isConfirmed"] = false
-        assign_attributes(attributes)
-        self.organizers |= competition_to_clone.organizers
-        self.delegates |= competition_to_clone.delegates
-      else
-        errors.add(:competition_id_to_clone, "invalid")
-      end
-    end
+  def build_clone
+    # Don't clone the following attributes.
+    attributes_to_clone = attributes
+    %w(id name cellName year month day endMonth endDay registration_open registration_close results_posted_at).each { |attribute| attributes_to_clone.delete attribute }
+    clone = Competition.new
+    clone.assign_attributes(attributes_to_clone)
+    clone.organizers = organizers
+    clone.delegates = delegates
+    clone.showAtAll = false
+    clone.isConfirmed = false
+    clone
```

Comment:
Yes, that really needs cleaning up!

I'd be a bit more ambitious and define both clonable and unclonable attributes. That makes this test possible...

`expect(Competition.column_names()).to match_array(Competition::CLONABLE_ATTRS + Competition::UNCLONABLE_ATTRS)`

..., which makes the clone code self maintaining! In the sense that if a column is added or renamed, the test will go Red, and we have to explicitly declare how the new column is treated.


Also, `showAtAll` and `isConfirmed` are set to `false` by the schema, so it seems wrong to set it here too.